sysfs-led-vince.c

```diff
@@ -0,0 +1,372 @@
+} led_paths_vince_t;
+
+typedef struct
+{
+    sysfsval_t *cached_max_brightness;
+    sysfsval_t *cached_brightness;
+    sysfsval_t *cached_blink;
+
+    int         control_value;
+    bool        control_blink;
+
+} led_channel_vince_t;
+
+/* ------------------------------------------------------------------------- *
+ * ONE_CHANNEL
+ * ------------------------------------------------------------------------- */
+
+static void        led_channel_vince_init          (led_channel_vince_t *self);
+static void        led_channel_vince_close         (led_channel_vince_t *self);
+static bool        led_channel_vince_probe         (led_channel_vince_t *self, const led_paths_vince_t *path);
+static void        led_channel_vince_set_value     (led_channel_vince_t *self, int value);
+static void        led_channel_vince_set_blink     (led_channel_vince_t *self, int on_ms, int off_ms);
+
+/* ------------------------------------------------------------------------- *
+ * ALL_CHANNELS
+ * ------------------------------------------------------------------------- */
+
+static void        led_control_vince_blink_cb      (void *data, int on_ms, int off_ms);
+static void        led_control_vince_value_cb      (void *data, int r, int g, int b);
+static void        led_control_vince_close_cb      (void *data);
+
+bool               led_control_vince_probe         (led_control_t *self);
+
+/* ========================================================================= *
+ * ONE_CHANNEL
+ * ========================================================================= */
+
+static void
+led_channel_vince_init(led_channel_vince_t *self)
+{
+    self->cached_max_brightness = sysfsval_create();
+    self->cached_brightness     = sysfsval_create();
+    self->cached_blink          = sysfsval_create();
+
+    self->control_blink = false;
+    self->control_value = 0;
+}
+
+static void
+led_channel_vince_close(led_channel_vince_t *self)
+{
+    sysfsval_delete(self->cached_max_brightness),
+        self->cached_max_brightness = 0;
+
+    sysfsval_delete(self->cached_brightness),
+        self->cached_brightness = 0;
+
+    sysfsval_delete(self->cached_blink),
+        self->cached_blink = 0;
+}
+
+static bool
+led_channel_vince_probe(led_channel_vince_t *self,
+                        const led_paths_vince_t *path)
+{
+    bool ack = false;
+
+    /* Probe control files in reverse existance likelihood order.
+     * Practically all led control directories have 'brightness'
+     * file, most have 'max_brightness' while only some have 'blink'.
+     */
+    if( !sysfsval_open_rw(self->cached_blink, path->blink) )
+        goto cleanup;
+
+    if( !sysfsval_open_ro(self->cached_max_brightness, path->max_brightness) )
+        goto cleanup;
+
+    /* The 'max_brightness' seems to be dynamic. Make an attempt
+     * to set it to an artificially high value and assume that kernel
+     * side will cap the value to the true maximum... */
+#if 1 // TODO: make a QUIRK out of this
```

Comment:
Would f5121 backend work if you disable this bit?

If yes, I'd rather handle the difference via configuration than by adding completely new backend.

Reply:
Can we handle `4096 level brightness` via config?

Reply:
Not in the current state, but... IIRC xperiax reported odd / lower than expected max brightness values which was causing issues and the q'n'd solution was to force the value that is used. And that should have come from configuration instead of hard-coding a workaround. So the solution I'm proposing would be: remove the hard-coded xperiax workaround -> the code should be able to handle vince too [with appropriate config files].

I could try to make the required code changes but as I do not have access to vince hw, I would need some testing help from you.

Reply:
Sure, it's my pleasure.

Reply:
Took a while, sorry about that.

Can you check whether this works for you: https://github.com/mer-hybris/mce-plugin-libhybris/pull/43

Reply:
It almost works, except the max brightness, every reboot reset the max_brightness to 150.

I have already set the `MaxBrightnessOverride`, seems it's not working well.


```
[root@Redmi5Plus mce]# cat 60-vince.ini
[LEDConfigHybris]

# Choose vince backend
BackEnd=f5121


RedDirectory=/sys/class/leds/red
GreenDirectory=/sys/class/leds/red
BlueDirectory=/sys/class/leds/red

MaxBrightnessOverride=255
RedMaxBrightnessOverride=255
GreenMaxBrightnessOverride=255
BlueMaxBrightnessOverride=255
# Built-in defaults for directory relative paths
#BrightnessFile=brightness
#MaxBrightnessFile=max_brightness

# Optional channel specific overrides
#RedBrightnessFile=/sys/class/leds/red/brightness
#RedMaxBrightnessFile=/sys/class/leds/red/max_brightness
# ... and similarly for Green.

[root@Redmi5Plus mce]# cat /sys/class/leds/red/max_brightness
150
[root@Redmi5Plus mce]# cat /sys/class/leds/red/brightness
25

```

Reply:
> [root@Redmi5Plus mce]# cat /sys/class/leds/red/max_brightness
> 150

Note that expectation is that content of this sysfs file is not changed.

But, now that I wrote that... I wonder if it should be  changed and whether current behavior has wandered off from original f5121 behavior? Need to take another look in XperiaX too...

I was assuming that vince has sensible value in max_brightness and it got ignored due to that f5121 hack. Now I wonder if vince too has too low max_brightness values that need to be overridden (also by writing to sysfs)?

Reply:
I made a tweak in the PR, basically: the configured value is written to sysfs max_brightness file

(Originally max_brightness value was written to, but situation changed later on as part of general changes and in case of f5121 that change probably was incorrect but went unnoticed)

Couple of notes

> MaxBrightnessOverride=255

This is enough if you want to have the same max brightness for all channels.

> RedMaxBrightnessOverride=255
> GreenMaxBrightnessOverride=255
> BlueMaxBrightnessOverride=255

These are needed in case you want to use different values for red / green / blue channels.

For testing purposes you could run mce manually to get more visibility on what is happening, e.g. disable normal logging but enable everything from led backend related parts

```
systemctl stop mce
mce -Tq -lsysfs*:*

```